libhb: support multi-pass with VP9 Constant Quality mode

samhutchins opened the pull request:
**Description of Change:**

This change adds support for multi-pass encoding with VP9 in CQ mode. There's an enhancement request open for this here: #1905

~I've only done this for the CLI in this PR as a) I don't immediately know how to do it in the various GUIs and b) I'm not sure if this will be accepted, or if the way I've done it makes sense. I'm happy to have a crack at the GUI changes necessary, but I'd probably need some guidance on where to look.~

Edit: I've made an attempt at the Mac and Linux GUIs, but I'm afraid I don't have access to a Windows box

I do feel like the change makes the code a little repetitive in `preset.c`, would it be better to find a way to do this without setting the multi-pass stuff in every branch of the if statements? I also wasn't sure if I could/should change the signature of `hb_video_multipass_is_supported` so it could handle both cases (ABR and CQ), rather than adding a new method for CQ mode.


**Tested on:**

- [ ] Windows 10+  (via MinGW)
- [x] macOS 10.13+
- [x] Ubuntu Linux

samhutchins committed:
libhb: support multi-pass with VP9 Constant Quality mode

 - Add a new method in common.c to check if multi-pass is supported for
   constant quality with a given codec
 - Change the CLI to set VideoMultiPass regardless of video quality type
 - Propagate constant-quality multi-pass all the way through

galad87 commented:
This is a bit of a minefield, because there are presets set to use CQ but with multi-pass enabled, and presets that can fallback to ABR.
Maybe it will be better to add a separate preset key. Or check whether VP9 CQ 2-pass does provide enough improvement to actually bother. It should be the only encoder that offers such option. SVT-AV1 had it too but it was removed a year ago.

samhutchins commented:
Was CQ 2-pass ever enabled with SVT_AV1 in HandBrake? I looked through the history for `hb_job_setup_passes` in `hb.c`, and it looks like multi-pass has always been disabled in constant quality mode, regardless of the encoder/preset being used:

```c
    if (job->vquality > HB_INVALID_VIDEO_QUALITY)
    {
        job->multipass = 0;
    }
```

It's been that way for a number of years. I believe my change only alters this behaviour for VP9.

I've had a look at at `preset_builtin.json` and I can see that all the svt_av1 presets have `VideoQualityType` set to 2 (which I understand to mean constant quality) and `VideoMultiPass` set to true. I've checked these presets with my patch applied and it still falls back to single pass, as expected.

I'm afraid I'm not sure I quite understand what you mean by presets falling back to ABR though.

I can see that this is a bit of a minefield with the presets. I don't really use HandBrake's presets, so I hadn't really considered their behaviour in this case. Thank you for pointing it out.

I did a test with BBB and saw a 200kbps bitrate savings with multi-pass compared to single pass, and I think that change would add up with longer videos. I can't tell the difference between the two versions by eye, but I'm not sure if you guys have an accepted quality metric I should be measuring with.

I'll leave whether or not it's worth it up to you, I had fun working out how to make this change in any case :-)

samhutchins committed:
MacGui: support multi-pass CQ with VP9

samhutchins commented:
I've had a go at the Mac gui to satisfy my own curiosity

samhutchins committed:
LinGui: validate encoder multipass support and only show turbo for encoders that support it

samhutchins committed:
LinGui: support multi-pass CQ with VP9

galad87 commented:
You could bump the presets major version, and add a function to disable multi-pass if the codec is VP9 and if quality is set to constant quality, so that people with existing presets won't automatically get multi-pass turned on.

Constant quality multi-pass has never been implemented for any codec, SVT-AV1 didn't have it.

samhutchins committed:
libhb: bump preset version, and disable multi-pass for existing CQ VP9 presets

 - We don't want people's existing presets to dramatically change in behaviour

samhutchins commented:
Something like this? [029a9c9](https://github.com/HandBrake/HandBrake/pull/5899/commits/029a9c95a90da837852eff247714bf8ac1aab284)

I'm having difficulty fully testing this, it works fine when importing an external json file into the presets, but I'm struggling to set up a test scenario for a user preset that's saved into the UserPresets file. I'm testing on a Mac, and when I rev the preset major version _some_ of my saved presets no longer show up in the presets window, and I'm not sure why. I'll keep trying to work it out though

I also noticed something that might be a bug, `import_44_0_0()` (and earlier) call the next `import_xx_x_x()` function at the end, but `import_47_0_0()` and onwards don't.

galad87 commented:
Looks like a bug, I guess I forgot that part 😄

HandBrake saves a copy of the previous presets in ~/Library/Containers/fr.handbrake.HandBrake/Data/Library/Application Support/HandBrake, so to test again you can delete the latest json file and rename the previous.

galad87 commented:
I fixed the missing import_xx_x_x() calls, it would help if you could rebase the PR.

samhutchins commented:
Rebased, and I've tested the preset upgrade from 1.7.3 which seems to work as expected. Thanks for the tip on where the preset files are stored :-)

galad87 commented on macosx/HBVideo+UIAdditions.m:
```diff
@@ -100,7 +100,7 @@ - (BOOL)fastDecodeSupported
 + (NSSet<NSString *> *)keyPathsForValuesAffectingTurboMultiPassSupported
 {
-    return [NSSet setWithObjects:@"encoder", nil];
+    return [NSSet setWithObjects:@"encoder", @"job.video.qualityType", nil];
```
@"job.video.qualityType" can be simplified to @"qualityType"

galad87 commented on macosx/HBVideo+UIAdditions.m:
```diff
@@ -111,12 +111,19 @@ - (BOOL)turboMultiPassSupported
 + (NSSet<NSString *> *)keyPathsForValuesAffectingMultiPassSupported
 {
-    return [NSSet setWithObjects:@"encoder", nil];
+    return [NSSet setWithObjects:@"encoder", @"job.video.qualityType", nil];
```
Here too.

galad87 commented:
Looks good, all that's left is to hook up the checkbox in the WinUI. I'll try soon.

galad87 committed:
WinGui: allow multipass for CQ when available.

galad87 commented:
This should be the changes for Windows: https://gist.github.com/galad87/070824b86d83c54a1863a53591be4d67

galad87 commented:
Like you wrote, making a single hb_video_multipass_is_supported() function would simplify things.

samhutchins commented:
I've applied that patch in the branch, thanks!

Would you like me to look into adjusting `hb_video_multipass_is_supported()` in this PR, or would you prefer a separate PR for that?

sr55 commented on win/CS/HandBrakeWPF/Services/Encode/Factories/EncodeTaskFactory.cs:
```diff
@@ -251,10 +251,10 @@ private Video CreateVideo(EncodeTask job)
-                video.Turbo = job.TurboAnalysisPass;
             }
 
+            video.MultiPass = job.MultiPass;
```
I might be inclined to special case guard this as ABR or VP9.

This likely protects against the UI state being ticked while disabled.  I'm not 100% sure it will actually be unticked when the checkbox is hidden.

samhutchins replied:
The Windows UI would be the only one to make that extra check for VP9, which would add an extra place to check/modify if this feature was enabled for any other encoders in the future. I know that VP8 also supports it, for example, and as it currently stands it would require no extra UI work to enable it (although my understanding is that it's considered deprecated by the HandBrake team, so I'm not expecting it to be enabled)

Additionally, it's my understanding that the job is further sanitised by libhb, which makes this same check in `hb.c` (https://github.com/HandBrake/HandBrake/pull/5899/files/940aba31888bb759ad9780961322458fc431e03f#diff-574ebf6c5d95e912886b84976f946f09a74d7c04c72be275182bafff202f14eaR1904)

That said, I'm unfamiliar with how the Windows UI works with respect to that checkbox being ticked in all situations, so as far as I'm concerned what you say goes. I'm still unable to test any changes on Windows myself, I'm afraid, but I can certainly add an extra `if` statement if required

galad87 replied:
Yes, libhb will sanitize the values. But if you want to checkbox to be unticked too when switching to constant quality if the encoder doesn't support 2-pass we can make it so.

sr55 replied:
Ok, cool. That's fine then.

samhutchins committed:
libhb: refactor to combine both supports_multipass methods into one

samhutchins commented:
I decided to do the refactor to only have one `hb_video_multipass_is_supported` function. I was able to get a Windows 10 VM set up, so I've been able to test that platform as well.

I think it simplifies the code in most places

galad87 commented:
Looks good. Pinging @robxnano for the GTK part.

Disabled checkbox on Windows are still colored, so the different state is not as noticeable as on Mac, but we can always fix that later.

galad87 commented:
Just thought that we could simply hide the multipass checkbox if it's not applicable, to avoid the issue of it being shown disabled and checked.

samhutchins commented:
Works for me. I've put together a rough prototype locally, I think it feels ok. I'm not sure if me continuing to add commits to this PR is helpful or not though, and I'm happy either way

robxnano commented on gtk/src/videohandler.c:
```diff
@@ -64,6 +64,21 @@ int ghb_set_video_preset(GhbValue *settings, int encoder, const char * preset)
+ghb_update_multipass(signal_user_data_t *ud)
+{
+    GtkWidget *multi_pass = GTK_WIDGET(ghb_builder_widget("VideoMultiPassBox"));
+    GtkWidget *turbo_multi_pass = GTK_WIDGET(ghb_builder_widget("VideoTurboMultiPass"));
```
There's no need to wrap these in GTK_WIDGET() macros as the function returns a GtkWidget pointer.

samhutchins replied:
Done

robxnano commented on gtk/src/videohandler.c:
```diff
@@ -64,6 +64,21 @@ int ghb_set_video_preset(GhbValue *settings, int encoder, const char * preset)
+    
+    bool constant_quality = ghb_dict_get_bool(ud->settings, "vquality_type_constant");
+    bool supports_multi_pass = hb_video_multipass_is_supported(encoder, constant_quality);
+    bool turbo_supported = (encoder & HB_VCODEC_X264_MASK) || (encoder & HB_VCODEC_X265_MASK);
```
For consistency with GLib and GTK APIs, can you change these bools to gboolean?

robxnano commented on gtk/src/callbacks.h:
```diff
@@ -85,5 +85,6 @@ void ghb_break_pts_duration(gint64 ptsDuration,
 void ghb_break_duration(gint64 duration, gint *hh, gint *mm, gint *ss);
 GtkFileFilter *ghb_add_file_filter(GtkFileChooser *chooser,
                                    const char *name, const char *id);
+void ghb_update_multipass(signal_user_data_t *ud);
```
Can you move this to videohandler.h? Since that's where the function definition is.

robxnano reviewed (changes requested):
Just a few minor issues I spotted. Is it OK if I request these changes? They're not serious enough to prevent the merge if the rest looks good.

samhutchins commented:
I'll make those changes, thanks!

samhutchins committed:
LinGui: cleanup

 - Remove unnecessary call to GTK_WIDGET
 - Use gboolean insteal of bool
 - Move declaration of ghb_update_multipass to correct header

bradleysepos commented:
@robxnano and @samhutchins You are both welcome and invited to join our developer channel #handbrake-dev on Libera.chat IRC, if you'd like.
